### ISSUE TITLE
Applying RNN

### ISSUE THREAD

**Issue by hpoit** [Andy-P/RecurrentNN.jl#8]
Could I receive some guidance on adjusting the net to correctly predict new targets from train.csv? 

https://s3-sa-east-1.amazonaws.com/nu-static/workable-data-science/data-science-puzzle.zip

**Comment by Andy-P** [comment]
@hpoit 
I would prefer not to to download and open zip files.

Perhaps if you provided a more specific question (ie. what are you trying to adjust? learning rate, size of net?) with some sample code and a sample of your data, I might be able to give you some guidance.  :)

**Comment by Andy-P** [comment]
Perhaps cutting and pasting a sample of your csv data would also be helpful

**Comment by hpoit** [comment]
Hi Andy, this is how train.csv looks like

```
julia> showcols(train)
24976x108 DataFrames.DataFrame
| Col # | Name        | Eltype     | Missing |
|-------|-------------|------------|---------|
| 1     | x016399044a | Int64      | 0       |
| 2     | x023c68873b | UTF8String | 0       |
| 3     | x0342faceb5 | Int64      | 0       |
| 4     | x04e7268385 | Int64      | 0       |
| 5     | x06888ceac9 | Int64      | 0       |
| 6     | x072b7e8f27 | Float64    | 0       |
| 7     | x087235d61e | Int64      | 0       |
| 8     | x0b846350ef | Float64    | 0       |
| 9     | x0e2ab0831c | Float64    | 0       |
| 10    | x12eda2d982 | Float64    | 0       |
| 11    | x136c1727c3 | Float64    | 0       |
| 12    | x173b6590ae | Float64    | 0       |
| 13    | x174825d438 | Int64      | 0       |
| 14    | x1f222e3669 | Float64    | 0       |
| 15    | x1f3058af83 | Int64      | 0       |
| 16    | x1fa099bb01 | Int64      | 0       |
| 17    | x20f1afc5c7 | Float64    | 0       |
| 18    | x253eb5ef11 | Float64    | 0       |
| 19    | x25bbf0e7e7 | Int64      | 0       |
| 20    | x2719b72c0d | Float64    | 0       |
| 21    | x298ed82b22 | Float64    | 0       |
| 22    | x29bbd86997 | Float64    | 0       |
| 23    | x2a457d15d9 | Int64      | 0       |
| 24    | x2bc6ab42f7 | Float64    | 0       |
| 25    | x2d7fe4693a | Float64    | 0       |
| 26    | x2e874bc151 | Float64    | 0       |
| 27    | x361f93f4d1 | UTF8String | 0       |
| 28    | x384bec5dd1 | Int64      | 0       |
| 29    | x3df2300fa2 | Float64    | 0       |
| 30    | x3e200bf766 | Int64      | 0       |
| 31    | x3eb53ae932 | Float64    | 0       |
| 32    | x435dec85e2 | Float64    | 0       |
| 33    | x4468394575 | Float64    | 0       |
| 34    | x49756d8e0f | Float64    | 0       |
| 35    | x4fc17427c8 | Float64    | 0       |
| 36    | x55907cc1de | Float64    | 0       |
| 37    | x55cf3f7627 | Float64    | 0       |
| 38    | x56371466d7 | Int64      | 0       |
| 39    | x5b862c0a8f | Int64      | 0       |
| 40    | x5f360995ef | Int64      | 0       |
| 41    | x60ec1426ce | Float64    | 0       |
| 42    | x63bcf89b1d | Float64    | 0       |
| 43    | x6516422788 | Float64    | 0       |
| 44    | x65aed7dc1f | Int64      | 0       |
| 45    | x6db53d265a | Int64      | 0       |
| 46    | x7734c0c22f | Float64    | 0       |
| 47    | x7743f273c2 | Float64    | 0       |
| 48    | x779d13189e | Float64    | 0       |
| 49    | x77b3b41efa | Float64    | 0       |
| 50    | x7841b6a5b1 | Float64    | 0       |
| 51    | x789b5244a9 | Float64    | 0       |
| 52    | x7925993f42 | Float64    | 0       |
| 53    | x7cb7913148 | Int64      | 0       |
| 54    | x7fe6cb4c98 | Float64    | 0       |
| 55    | x8311343404 | Float64    | 0       |
| 56    | x87b982928b | Float64    | 0       |
| 57    | x8a21502326 | Float64    | 0       |
| 58    | x8c2e088a3d | Int64      | 0       |
| 59    | x8d0606b150 | UTF8String | 0       |
| 60    | x8de0382f02 | Float64    | 0       |
| 61    | x8f5f7c556a | Int64      | 0       |
| 62    | x91145d159d | UTF8String | 0       |
| 63    | x96c30c7eef | Float64    | 0       |
| 64    | x96e6f0be58 | Float64    | 0       |
| 65    | x98475257f7 | Float64    | 0       |
| 66    | x99d44111c9 | Float64    | 0       |
| 67    | x9a575e82a4 | Int64      | 0       |
| 68    | x9b6e0b36c2 | Float64    | 0       |
| 69    | a14fd026ce  | Int64      | 0       |
| 70    | a24802caa5  | Float64    | 0       |
| 71    | aa69c802b6  | Float64    | 0       |
| 72    | abca7a848f  | Int64      | 0       |
| 73    | ac826f0013  | Float64    | 0       |
| 74    | ae08d2297e  | Int64      | 0       |
| 75    | aee1e4fc85  | Float64    | 0       |
| 76    | b4112a94a6  | Float64    | 0       |
| 77    | b709f75447  | Float64    | 0       |
| 78    | b835dfe10f  | UTF8String | 0       |
| 79    | b9a487ac3c  | Float64    | 0       |
| 80    | ba54a2a637  | Float64    | 0       |
| 81    | bdf934caa7  | Float64    | 0       |
| 82    | beb6e17af1  | Int64      | 0       |
| 83    | c0c3df65b1  | Int64      | 0       |
| 84    | c1b8ce2354  | Float64    | 0       |
| 85    | c58f611921  | Float64    | 0       |
| 86    | d035af6ffa  | Float64    | 0       |
| 87    | d2c775fa99  | Int64      | 0       |
| 88    | d4d6566f9c  | Float64    | 0       |
| 89    | dcfcbc2ea1  | Int64      | 0       |
| 90    | e0a0772df0  | Float64    | 0       |
| 91    | e16e640635  | UTF8String | 0       |
| 92    | e5efa4d39a  | Float64    | 0       |
| 93    | e7ee22effb  | Int64      | 0       |
| 94    | e86a2190c1  | Int64      | 0       |
| 95    | ea0f4a32e3  | Float64    | 0       |
| 96    | ed7e658a27  | Float64    | 0       |
| 97    | ee2ac696ff  | Float64    | 0       |
| 98    | f013b60e50  | Float64    | 0       |
| 99    | f0a0febd35  | Float64    | 0       |
| 100   | f1f0984934  | UTF8String | 0       |
| 101   | f66b98dd69  | Float64    | 0       |
| 102   | fbf66c8021  | Float64    | 0       |
| 103   | fdf8628ca7  | Int64      | 0       |
| 104   | fe0318e273  | Float64    | 0       |
| 105   | fe8cdd80ba  | Float64    | 0       |
| 106   | ffd1cdcfc1  | Float64    | 0       |
| 107   | id          | Int64      | 0       |
| 108   | target      | Float64    | 0       |
```

I'm sorry I understand this isn't an issue about the package, but a question on how to use it. 

I have ran a hello world of Mocha.jl on MNIST but not yet for this one. My understanding is that RNNs use shared params over time (not space like CNNs) to extract patterns of sequences, and are proper for time series data, which I'm assuming this is for having originated from a bank. 

I would need some guidance on how to apply this package on this 108 column by 24976 row matrix of variables of no particular meaning. The UTF8String columns are SHA-256 hashed categorical variables. I am pretty sure each row represents a client, so 24976 clients at the time. The hashed values are probably client name, address, e-mail, telephone, etc. so I wouldn't worry about them.

**Comment by Andy-P** [comment]
@hpoit 

Generally, RNNs are used to predict the next variable(s) in a sequence. They are very useful for learning sequential sequence, but this data does not seem to be sequential.  This case seems better suited to a standard DNN, or a perhaps a CNN.

Also note the "Hello World"  of RNN is generally a char-RNN that learns the next character in a sentence. If you are just starting, I suggest having a look at Karpathy"s online example. 
[http://cs.stanford.edu/people/karpathy/recurrentjs/](http://cs.stanford.edu/people/karpathy/recurrentjs/)

**Comment by hpoit** [comment]
Thanks Andy

**Comment by hpoit** [comment]
Andy check these out

http://cs229.stanford.edu/proj2012/BernalFokPidaparthi-FinancialMarketTimeSeriesPredictionwithRecurrentNeural.pdf

http://deeplearning.cs.cmu.edu/notes/shaoweiwang.pdf

http://stats.stackexchange.com/questions/127542/convolutional-neural-network-for-time-series

Maybe we could work together on this one. I haven't found any RNN ready in Julia for time series. My focus isn't financial, but anything that is worth being observed through time.
@johnmyleswhite @StefanKarpinski

**Comment by hpoit** [comment]
`| Row   | fbf66c8021 | fdf8628ca7 | fe0318e273  | fe8cdd80ba | ffd1cdcfc1 | id    | target     |
|-------|------------|------------|-------------|------------|------------|-------|------------|
| 1     | 91.7786    | 1058       | 1.0005      | 40.2978    | 1.74053    | 0     | 0.536383   |
| 2     | 26.235     | 1036       | 1.0006      | 6.03076    | 7.08524    | 1     | 0.426536   |
| 3     | 48.9608    | 1039       | 0.000571041 | 40.2693    | 5.31598    | 2     | 0.498649   |
| 4     | 6.23632    | 1044       | 1.0007      | 29.1151    | 1.78314    | 3     | 1.01357    |
| 5     | 19.3486    | 998        | 1.00058     | 44.0475    | 3.21903    | 4     | 4.00876    |
| 6     | 62.7932    | 988        | 1.00053     | 43.2183    | 8.44368    | 5     | 1.77013    |
| 7     | 97.5266    | 1049       | 1.00069     | 10.2458    | 7.99116    | 6     | 0.895514   |
| 8     | 62.2518    | 1025       | 1.00073     | 36.1519    | 3.99028    | 7     | 0.348059   |
| 9     | 29.6062    | 1013       | 1.00059     | 13.0537    | 1.14704    | 8     | 0.547179   |
| 10    | 1.69667    | 1010       | 1.00059     | 2.71338    | 2.28359    | 9     | 2.38998    |
| 11    | 28.5268    | 970        | 1.00049     | 38.6509    | 2.38913    | 10    | 3.36091    |
| 12    | 83.2608    | 985        | 1.00078     | 38.2881    | 1.14008    | 11    | 1.4124     |
| 13    | 91.1006    | 976        | 1.00076     | 15.5142    | 4.79976    | 12    | -0.0996988 |
| 14    | 11.0477    | 981        | 1.00089     | 23.5938    | 3.45635    | 13    | 0.40257    |
| 15    | 65.0404    | 969        | 1.00069     | 42.4824    | 3.15091    | 14    | 0.896381   |
| 16    | 40.3231    | 1025       | 1.00035     | 39.045     | 1.12076    | 15    | 1.97511    |
| 17    | -3.04205   | 1008       | 1.00069     | 14.4767    | 4.3642     | 16    | 0.315973   |
| 18    | -19.6236   | 1017       | 1.0006      | 32.9957    | 1.78944    | 17    | 1.35555    |
| 19    | 8.65733    | 954        | 1.00044     | 45.9944    | 2.35002    | 18    | 5.32482    |
| 20    | 31.6925    | 1046       | 1.00055     | 39.5133    | 2.96461    | 19    | 0.591043   |
| 21    | 81.1518    | 979        | 1.00049     | 18.877     | 3.95433    | 20    | 0.610452   |
| 22    | 17.8116    | 963        | 1.00045     | 44.2147    | 2.2464     | 21    | 1.45285    |
| 23    | 44.6326    | 1006       | 1.00049     | 33.1826    | 2.8726     | 22    | 0.179439   |
⋮
| 24953 | 67.7065    | 966        | 1.00046     | 45.3271    | 3.86463    | 24952 | 7.6247     |
| 24954 | 48.4608    | 925        | 0.000645519 | 10.0123    | 5.18167    | 24953 | 0.340788   |
| 24955 | 15.8755    | 1009       | 1.00075     | 4.44004    | 3.71273    | 24954 | 0.205181   |
| 24956 | 66.2357    | 982        | 1.00049     | 40.2374    | 3.73331    | 24955 | 2.04826    |
| 24957 | 7.30888    | 1048       | 1.00046     | 48.0095    | 2.96435    | 24956 | 0.150869   |
| 24958 | 38.1056    | 1027       | 1.00057     | 34.4642    | 2.89719    | 24957 | 0.775302   |
| 24959 | 54.3954    | 959        | 1.00068     | 43.4772    | 2.30724    | 24958 | 0.214532   |
| 24960 | 53.4419    | 1013       | 1.00057     | 8.44832    | 2.55465    | 24959 | 0.1074     |
| 24961 | 54.858     | 1031       | 1.00056     | 4.92166    | 1.29723    | 24960 | 0.46036    |
| 24962 | 2.46157    | 987        | 1.00056     | 24.7661    | 5.32419    | 24961 | 0.241159   |
| 24963 | -1.48726   | 1034       | 1.00047     | 43.5869    | 1.45187    | 24962 | 1.73751    |
| 24964 | 64.2474    | 1024       | 1.00052     | 25.3957    | 2.0713     | 24963 | 0.390916   |
| 24965 | 24.4898    | 1008       | 1.00041     | 9.17904    | 2.66892    | 24964 | 3.03904    |
| 24966 | 1.82829    | 985        | 1.0008      | 23.6407    | 2.76205    | 24965 | 2.71701    |
| 24967 | 8.61764    | 981        | 1.00067     | 33.3696    | 4.66886    | 24966 | 2.94019    |
| 24968 | 18.3361    | 1024       | 1.00062     | 16.4426    | 2.93404    | 24967 | 0.309083   |
| 24969 | 86.9393    | 982        | 1.00046     | 8.75922    | 4.45587    | 24968 | 0.998051   |
| 24970 | 84.2475    | 980        | 1.00068     | 6.66636    | 1.15699    | 24969 | 0.762129   |
| 24971 | 20.744     | 980        | 1.00059     | 24.2177    | 2.89352    | 24970 | 0.377102   |
| 24972 | 9.25122    | 998        | 1.00051     | 48.5923    | 0.837068   | 24971 | 3.88774    |
| 24973 | 56.7977    | 1006       | 1.00029     | 20.9637    | 1.36332    | 24972 | 2.39431    |
| 24974 | -41.5304   | 954        | 1.00045     | 19.2523    | 1.42232    | 24973 | 0.50462    |
| 24975 | 40.2701    | 1035       | 1.00067     | 43.852     | 1.82505    | 24974 | 23.0103    |
| 24976 | 12.7766    | 1043       | 1.00066     | 7.67478    | 1.65179    | 24975 | 2.53425    |`

**Comment by hpoit** [comment]
This is the target column

`julia> train[108]
24976-element DataArrays.DataArray{Float64,1}:
  0.536383 
  0.426536 
  0.498649 
  1.01357  
  4.00876  
  1.77013  
  0.895514 
  0.348059 
  0.547179 
  2.38998  
  3.36091  
  1.4124   
 -0.0996988
  0.40257  
  0.896381 
  1.97511  
  0.315973 
  1.35555  
  5.32482  
  0.591043 
  0.610452 
  1.45285  
  0.179439 
  3.16545  
  ⋮        
  7.6247   
  0.340788 
  0.205181 
  2.04826  
  0.150869 
  0.775302 
  0.214532 
  0.1074   
  0.46036  
  0.241159 
  1.73751  
  0.390916 
  3.03904  
  2.71701  
  2.94019  
  0.309083 
  0.998051 
  0.762129 
  0.377102 
  3.88774  
  2.39431  
  0.50462  
 23.0103   
  2.53425`

... which appears to be the output of a logit. This can be done on a CNN, sure, but with no use of shared params over time (memory).

**Comment by hpoit** [comment]
I'm also comparing your rnn.jl code to https://github.com/kzahedi/RNN.jl/blob/master/src/RNN.jl and trying to understand the differences.

**Comment by hpoit** [comment]
If by sequential sequence you meant sampling for the next prediction every epoch (very greedy), there's a more sophisticated way to do this that returns better results, which is by sampling multiple times per epoch, so single bad choices permanently recorded are prevented. This should enable full sequence prediction, which I will look into.

**Comment by hpoit** [comment]
> Sequences. Depending on your background you might be wondering: What makes Recurrent Networks so special? A glaring limitation of Vanilla Neural Networks (and also Convolutional Networks) is that their API is too constrained: they accept a fixed-sized vector as input (e.g. an image) and produce a fixed-sized vector as output (e.g. probabilities of different classes). Not only that: These models perform this mapping using a fixed amount of computational steps (e.g. the number of layers in the model). The core reason that recurrent nets are more exciting is that they allow us to operate over sequences of vectors: Sequences in the input, the output, or in the most general case both. 
http://karpathy.github.io/2015/05/21/rnn-effectiveness/

**Comment by hpoit** [comment]
> Sequences. Depending on your background you might be wondering: What makes Recurrent Networks so special? A glaring limitation of Vanilla Neural Networks (and also Convolutional Networks) is that their API is too constrained: they accept a fixed-sized vector as input (e.g. an image) and produce a fixed-sized vector as output (e.g. probabilities of different classes). Not only that: These models perform this mapping using a fixed amount of computational steps (e.g. the number of layers in the model). The core reason that recurrent nets are more exciting is that they allow us to operate over sequences of vectors: Sequences in the input, the output, or in the most general case both. 

http://karpathy.github.io/2015/05/21/rnn-effectiveness/

https://cs.stanford.edu/~quocle/tutorial2.pdf

**Comment by hpoit** [comment]
@johnmyleswhite @StefanKarpinski am I making sense?

**Comment by pazzo83** [comment]
@Andy-P 
I have an additional question about usage.  How would you apply this library to doing time-series prediction, as in the link below?
http://machinelearningmastery.com/time-series-prediction-lstm-recurrent-neural-networks-python-keras/

I guess one of the main things I don't quite grasp in your example is the usage of the "wil" NNMatrix and how that would be used outside of the example (if it would even be necessary).

Thanks!

Chris

**Comment by Andy-P** [comment]
@pazzo83 This library is longer supported. I suggest you use [MxNet.jl](http://dmlc.ml/MXNet.jl/). There is a [good example of an LSTM here](http://dmlc.ml/MXNet.jl/latest/tutorial/char-lstm/)
